py/dml/codegen.py

```diff
@@ -1306,18 +1306,7 @@ def expr_cast(tree, location, scope):
     for (site, _) in struct_defs:
         report(EANONSTRUCT(site, "'cast' expression"))
 
-    if (compat.dml12_misc in dml.globals.enabled_compat
-        and isinstance(expr, InterfaceMethodRef)):
-        # Workaround for SIMICS-9868
-        return mkLit(tree.site, "%s->%s" % (
-            expr.node_expr.read(), expr.method_name), type)
-
-    if isinstance(expr, NonValue) and (
-            not isinstance(expr, NodeRef)
-            or not isinstance(safe_realtype(type), TTrait)):
-        raise expr.exc()
-    else:
-        return mkCast(tree.site, expr, type)
+    return mkCast(tree.site, expr, type)
```

Comment:
There is a convention that you should check `NonValue` explicitly after `codegen_expression_maybe_nonvalue`, so it's possibly better to keep the `if` where it was, even if it's functionally redundant.

Reply:
Alternatively, we could break out the non-value cases of mkCast to separate functions.

Reply:
Discussed verbally